libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -408,11 +408,20 @@ open class GossipRouter(
                     .plus(getDirectPeers())
             } else {
                 msg.topics
-                    .mapNotNull { topic ->
-                        mesh[topic] ?: fanout[topic] ?: getTopicPeers(topic).shuffled(random).take(params.D)
-                            .also {
-                                if (it.isNotEmpty()) fanout[topic] = it.toMutableSet()
-                            }
+                    .map { topic ->
+                        val topicMeshPeers = mesh[topic]
+                        if (topicMeshPeers != null) {
+                            // we are subscribed to the topic
+                            val addFromNonMeshCount = max(0, params.D - topicMeshPeers.size)
+                            val nonMeshTopicPeers = (getTopicPeers(topic) - topicMeshPeers)
```

Comment:
Do we care about the score of the topic peers in this case?

Reply:
> Do we care about the score of the topic peers in this case?

I think the scoring is already "captured" by the mesh. Outside that, seems reasonable to me to pick random peers. 

On Friday I had a similar comment suggesting "why don't we select by sorting them by score" and Anton had a good reply to it.

Reply:
Yeah, good question. Here is the [quotation](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#score-thresholds) from the spec: 
> `PublishThreshold`: when a peer's score drops below this threshold, self-published messages are not propagated towards this peer when (flood) publishing. This threshold should be negative, and less than or equal to the gossip threshold.

I believe this mean that both 'flood' and 'regular' publishing should consider this threshold. 
It's probably counter eclipse attack measure. 
So I believe we still need to truncate those with the score `< PublishThreshold` _but_ still consider them when there are few peers to publish

Does it make sense?

Reply:
Yes that seems like a convincing interpretation to me.